File: android/Comandas/app/src/main/java/com/valleapp/comandas/OpMesas.java
package com.valleapp.comandas;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.ComponentName;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Color;
import android.os.Bundle;
import android.os.IBinder;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;

import com.valleapp.comandas.db.DBCuenta;
import com.valleapp.comandas.db.DBMesas;
import com.valleapp.comandas.db.DBZonas;
import com.valleapp.comandas.utilidades.Instruccion;
import com.valleapp.comandas.utilidades.ServicioCom;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;


public class OpMesas extends Activity {

    DBZonas dbZonas;
    DBMesas dbMesas;
    DBCuenta dbCuenta;
    JSONObject mesa ;
    String op ;
    JSONArray lsmesas = null;
    JSONArray lszonas = null;
    JSONObject zn = null;
    JSONObject art = null;
    private String url;
    Context cx;
    private String server;
    private ServicioCom servicioCom;

    private final ServiceConnection mConexion = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            servicioCom = ((ServicioCom.MyBinder)iBinder).getService();
            if (servicioCom != null){
                dbZonas = (DBZonas) servicioCom.getDb("zonas");
                dbMesas = (DBMesas) servicioCom.getDb("mesas");
                dbCuenta = (DBCuenta) servicioCom.getDb("lineaspedido");
                rellenarZonas();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            servicioCom = null;
        }
    };



    private void rellenarZonas() {
        try {

            lszonas = dbZonas.getAll();

            if(lszonas.length()>0){

                LinearLayout ll = findViewById(R.id.pneZonas);
                ll.removeAllViews();
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,LinearLayout.LayoutParams.MATCH_PARENT);

                params.setMargins(5,0,5,0);
                for (int i = 0; i < lszonas.length(); i++) {
                    JSONObject  z =  lszonas.getJSONObject(i);

                    if(zn==null && i==0) zn=z;

                    LayoutInflater inflater = (LayoutInflater)cx.getSystemService
                            (Context.LAYOUT_INFLATER_SERVICE);
                    View v = inflater.inflate(R.layout.btn_art, null);


                    Button btn = v.findViewById(R.id.boton_art);
                    btn.setId(i);
                    btn.setSingleLine(false);
                    btn.setTextSize(11);
                    btn.setText(z.getString("Nombre").trim().replace(" ", "\n"));
                    String[] rgb = z.getString("RGB").trim().split(",");
                    btn.setBackgroundColor(Color.rgb(Integer.parseInt(rgb[0].trim()), Integer.parseInt(rgb[1].trim()), Integer.parseInt(rgb[2].trim())));

                    btn.setOnClickListener(view -> {
                        try {
                            zn = lszonas.getJSONObject(view.getId());
                            rellenarMesas();
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                    });
                    ll.addView(v, params);


                }

                rellenarMesas();
            }

        }catch (Exception e){
            e.printStackTrace();
        }

    }

    private void rellenarMesas() {
        try {


            lsmesas = dbMesas.getAll(zn.getString("ID"));

            if(lsmesas.length()>0){

                TableLayout ll = findViewById(R.id.pneMesas);
                ll.removeAllViews();
                TableLayout.LayoutParams params = new TableLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT);

                DisplayMetrics metrics = getResources().getDisplayMetrics();
                TableRow.LayoutParams rowparams = new TableRow.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT, Math.round(metrics.density * 120));

                rowparams.setMargins(5,5,5,5);

                TableRow row = new TableRow(cx);
                ll.addView(row, params);

                for (int i = 0; i < lsmesas.length(); i++) {
                    JSONObject m_show = lsmesas.getJSONObject(i);

                    LayoutInflater inflater = (LayoutInflater) cx.getSystemService
                            (Context.LAYOUT_INFLATER_SERVICE);

                    @SuppressLint("InflateParams") View v = inflater.inflate(R.layout.btn_art, null);


                    Button btn = v.findViewById(R.id.boton_art);

                    btn.setId(i);
                    btn.setSingleLine(false);
                    btn.setText(m_show.getString("Nombre"));
                    btn.setTag(m_show);
                    btn.setTextSize(15);
                    String[] rgb = m_show.getString("RGB").trim().split(",");
                    if (m_show.getInt("ID") != mesa.getInt("ID")) {
                        btn.setBackgroundColor(Color.rgb(Integer.parseInt(rgb[0].trim()),
                                Integer.parseInt(rgb[1].trim()),
                                Integer.parseInt(rgb[2].trim())));
                        btn.setOnClickListener(view -> {
                            try {
                                JSONObject m = (JSONObject) view.getTag();
                                ContentValues p = new ContentValues();

                                if (art == null) {
                                    p.put("idp", mesa.getString("ID"));
                                    p.put("ids", m.getString("ID"));
                                    finalizar(m);
                                } else {
                                    String idm = m.getString("ID");
                                    String idLinea = art.getString("ID");
                                    p.put("idm", idm);
                                    p.put("idLinea", idLinea);
                                    dbCuenta.moverLinea(m, art);
                                }
                                if (servicioCom != null) {
                                    servicioCom.addColaInstrucciones(new Instruccion(p, url));
                                    finish();
                                }
                            } catch (Exception e) {
                                e.printStackTrace();
                            }

                        });
                    }else{
                        btn.setBackgroundResource(R.drawable.bg_pink);
                        btn.setOnClickListener(view ->{
                            Toast.makeText(cx,"Esta es la misma mesa..", Toast.LENGTH_SHORT).show();
                        });
                    }

                    row.addView(v, rowparams);

                    if (((i + 1) % 3) == 0) {
                        row = new TableRow(cx);
                        ll.addView(row, params);
                    }
                }


            }

        }catch (Exception e){
            e.printStackTrace();
        }

    }

    private void finalizar(JSONObject m) throws JSONException {
        if(op.equals("cambiar")){
            if(m.getString("abierta").equals("0")){
                dbMesas.abrirMesa(m.getString("ID"), mesa.getString("num"));
                dbMesas.cerrarMesa(mesa.getString("ID"));
                dbCuenta.cambiarCuenta(mesa.getString("ID"), m.getString("ID"));
            }else{
                dbCuenta.cambiarCuenta(mesa.getString("ID"), "-100");
                dbCuenta.cambiarCuenta(m.getString("ID"), mesa.getString("ID"));
                dbCuenta.cambiarCuenta("-100", m.getString("ID"));
            }
        }else {
            if(m.getString("abierta").equals("1")){
                dbMesas.cerrarMesa(m.getString("ID"));
                dbCuenta.cambiarCuenta(m.getString("ID"), mesa.getString("ID"));
            }else{
                dbMesas.abrirMesa(m.getString("ID"), mesa.getString("num"));
            }
        }

        Toast.makeText(getApplicationContext(), "Realizando un cambio en las mesas.....", Toast.LENGTH_SHORT).show();
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_op_mesas);
        server = getIntent().getExtras().getString("url");
        cx = this;
        op = getIntent().getExtras().getString("op");
        try {
            TextView l = findViewById(R.id.lblTitulo);
            String titulo = "";
            mesa = new JSONObject(getIntent().getExtras().getString("mesa"));
            if(op.equals("art")){
               titulo = "Cambiar articulo "+mesa.getString("Nombre");
               art = new JSONObject(getIntent().getExtras().getString("art"));
               url = server + "/cuenta/mvlinea";
            }else{
               url = server + (op.equals("cambiar") ? "/cuenta/cambiarmesas" :"/cuenta/juntarmesas");
               titulo =  (op.equals("cambiar") ? "Cambiar mesa " : "Juntar mesa ") + mesa.getString("Nombre");
            }
           l.setText(titulo);

        } catch (JSONException e) {
            e.printStackTrace();
        }


    }

    @Override
    protected void onResume() {
        Intent intent = new Intent(getApplicationContext(), ServicioCom.class);
        intent.putExtra("url", server);
        bindService(intent, mConexion, Context.BIND_AUTO_CREATE);
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        unbindService(mConexion);
        super.onDestroy();
    }
}

File: android/ValleTPV/ValleCOM/src/main/java/com/valleapp/vallecom/activitys/OpMesas.kt
package com.valleapp.vallecom.activitys


import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.ContentValues
import android.content.Intent
import android.content.ServiceConnection
import android.graphics.Color
import android.os.Bundle
import android.os.IBinder
import android.view.LayoutInflater
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import android.widget.Toast
import com.valleapp.vallecom.utilidades.ActivityBase
import com.valleapp.vallecom.utilidades.ServiceCOM
import com.valleapp.valletpv.R
import com.valleapp.valletpvlib.db.DBCuenta
import com.valleapp.valletpvlib.db.DBMesas
import com.valleapp.valletpvlib.db.DBZonas
import com.valleapp.valletpvlib.tools.Instruccion
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import kotlin.math.roundToInt
import com.valleapp.valletpvlib.R as LibR

class OpMesas : ActivityBase() {
    private lateinit var dbZonas: DBZonas
    private lateinit var dbMesas: DBMesas
    private lateinit var dbCuenta: DBCuenta
    private lateinit var mesa: JSONObject
    private lateinit var op: String
    private var lsmesas: JSONArray? = null
    private var lszonas: JSONArray? = null
    private var zn: JSONObject? = null
    private var artsToMove: JSONArray? = null // Usaremos este para guardar el array de artículos
    private lateinit var url: String
    private var servicioCom: ServiceCOM? = null

    private val mConexion: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            servicioCom = (iBinder as ServiceCOM.MyBinder).getService()
            if (servicioCom != null) {
                dbZonas = servicioCom!!.getDb("zonas") as DBZonas
                dbMesas = servicioCom!!.getDb("mesas") as DBMesas
                dbCuenta = servicioCom!!.getDb("lineaspedido") as DBCuenta
                rellenarZonas()
            }
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            servicioCom = null
        }
    }

    @SuppressLint("InflateParams")
    private fun rellenarZonas() {
        try {
            lszonas = dbZonas.getAll()
            if (lszonas!!.length() > 0) {
                val ll = findViewById<LinearLayout>(R.id.pneZonas)
                ll.removeAllViews()
                val params = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT
                )
                params.setMargins(5, 0, 5, 0)
                for (i in 0 until lszonas!!.length()) {
                    val z = lszonas!!.getJSONObject(i)
                    if (zn == null && i == 0) zn = z
                    val inflater = cx.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
                    val v = inflater.inflate(LibR.layout.btn_simple, null)
                    val btn = v.findViewById<Button>(LibR.id.btn_simple)
                    btn.id = i
                    btn.isSingleLine = false
                    btn.textSize = 11f
                    btn.text = z.getString("Nombre").trim { it <= ' ' }.replace(" ", "\n")
                    val rgb = z.getString("RGB").trim { it <= ' ' }.split(",").toTypedArray()
                    btn.setBackgroundColor(
                        Color.rgb(
                            rgb[0].trim { it <= ' ' }.toInt(),
                            rgb[1].trim { it <= ' ' }.toInt(),
                            rgb[2].trim { it <= ' ' }.toInt()
                        )
                    )
                    btn.setOnClickListener { view ->
                        try {
                            zn = lszonas!!.getJSONObject(view.id)
                            rellenarMesas()
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        }
                    }
                    ll.addView(v, params)
                }
                rellenarMesas()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun rellenarMesas() {
        try {
            lsmesas = dbMesas.getAll(zn!!.getString("ID"))
            if (lsmesas!!.length() > 0) {
                val ll = findViewById<TableLayout>(R.id.pneMesas)
                ll.removeAllViews()
                val params = TableLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
                val metrics = resources.displayMetrics
                val rowparams = TableRow.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, (metrics.density * 120).roundToInt()
                )
                rowparams.setMargins(5, 5, 5, 5)
                var row = TableRow(cx)
                ll.addView(row, params)
                for (i in 0 until lsmesas!!.length()) {
                    val mShow = lsmesas!!.getJSONObject(i)
                    val inflater = cx.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
                    val v = inflater.inflate(LibR.layout.btn_simple, ll, false)
                    val btn = v.findViewById<Button>(LibR.id.btn_simple)
                    btn.id = i
                    btn.isSingleLine = false
                    btn.text = mShow.getString("Nombre")
                    btn.tag = mShow
                    btn.textSize = 15f
                    val rgb = mShow.getString("RGB").trim { it <= ' ' }.split(",").toTypedArray()
                    if (mShow.getInt("ID") != mesa.getInt("ID")) {
                        btn.setBackgroundColor(
                            Color.rgb(
                                rgb[0].trim { it <= ' ' }.toInt(),
                                rgb[1].trim { it <= ' ' }.toInt(),
                                rgb[2].trim { it <= ' ' }.toInt()
                            )
                        )
                        btn.setOnClickListener { view ->
                            try {
                                val m = view.tag as JSONObject
                                val p = ContentValues()
                                // Aquí comprobamos si estamos moviendo artículos o mesas completas
                                if (op == "arts") { // Si la operación es para mover artículos (uno o más)
                                    val idm = m.getString("ID") // ID de la mesa destino
                                    val idLineas = mutableListOf<String>()

                                    // Iteramos sobre el array de artículos a mover
                                    artsToMove?.let {
                                        for (j in 0 until it.length()) {
                                            val obj = it.getJSONObject(j)
                                            val idLinea = obj.getString("ID")
                                            idLineas.add(idLinea)
                                            dbCuenta.cambiarLinea(idLinea, idm) // Mueve cada línea en la DB local
                                        }
                                    }
                                    p.put("idm", idm)
                                    p.put("idlineas", JSONArray(idLineas).toString()) // Enviamos el array de IDs de líneas
                                    url = "$server/cuenta/mvlineamultiple" // Usamos siempre la URL para múltiples líneas
                                } else { // Si la operación es para cambiar/juntar mesas completas
                                    p.put("idp", mesa.getString("ID"))
                                    p.put("ids", m.getString("ID"))
                                    finalizar(m) // Llama a la lógica de mesas completas
                                }

                                if (servicioCom != null) {
                                    servicioCom!!.agregarInstruccion(Instruccion(p, url))
                                    finish()
                                }
                            } catch (e: Exception) {
                                e.printStackTrace()
                            }
                        }
                    } else {
                        btn.setBackgroundResource(LibR.drawable.bg_pink)
                        btn.setOnClickListener {
                            Toast.makeText(cx, "Esta es la misma mesa..", Toast.LENGTH_SHORT).show()
                        }
                    }
                    row.addView(v, rowparams)
                    if ((i + 1) % 3 == 0) {
                        row = TableRow(cx)
                        ll.addView(row, params)
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @Throws(JSONException::class)
    private fun finalizar(m: JSONObject) {
        if (op == "cambiar") {
            if (m.getString("abierta") == "0") {
                dbMesas.abrirMesa(m.getString("ID"))
                dbMesas.cerrarMesa(mesa.getString("ID"))
                dbCuenta.cambiarCuenta(mesa.getString("ID"), m.getString("ID"))
            } else {
                dbCuenta.cambiarCuenta(mesa.getString("ID"), "-100")
                dbCuenta.cambiarCuenta(m.getString("ID"), mesa.getString("ID"))
                dbCuenta.cambiarCuenta("-100", m.getString("ID"))
            }
        } else {
            if (m.getString("abierta") == "1") {
                dbMesas.cerrarMesa(m.getString("ID"))
                dbCuenta.cambiarCuenta(m.getString("ID"), mesa.getString("ID"))
            } else {
                dbMesas.abrirMesa(m.getString("ID"))
            }
        }
        Toast.makeText(applicationContext, "Realizando un cambio en las mesas.....", Toast.LENGTH_SHORT).show()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_op_mesas)
        server = intent.extras!!.getString("url")!!
        op = intent.extras!!.getString("op")!!
        try {
            val l = findViewById<TextView>(R.id.title)
            var titulo = ""
            mesa = JSONObject(intent.extras!!.getString("mesa")!!)

            // Aquí se unifica la lógica para "art" y "artmultiple"
            if (op == "arts") { // Ahora se espera siempre un array de artículos
                titulo = "Cambiar artículo(s) " + mesa.getString("Nombre")
                // Siempre esperamos un JSONArray para 'artsToMove'
                // Si antes venía un solo JSONObject en "art", ahora debe venir envuelto en un JSONArray.
                // Si antes venía en "reg" para "artmultiple", ahora debe venir como "artsToMove"
                artsToMove = JSONArray(intent.extras!!.getString("arts")!!) // Cambiamos de "art" o "reg" a "arts"
                url = "$server/cuenta/mvlineamultiple" // Siempre usaremos la URL para múltiples líneas
            } else { // El resto de operaciones (cambiar/juntar mesas) quedan igual
                url = if (op == "cambiar") "$server/cuenta/cambiarmesas" else "$server/cuenta/juntarmesas"
                titulo = (if (op == "cambiar") "Cambiar mesa " else "Juntar mesa ") + mesa.getString("Nombre")
            }
            l.text = titulo
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    override fun onResume() {
        if (servicioCom == null) {
            val intent = Intent(applicationContext, ServiceCOM::class.java)
            intent.putExtra("url", server)
            bindService(intent, mConexion, BIND_AUTO_CREATE)
        } else {
            rellenarZonas()
        }
        super.onResume()
    }

    override fun onDestroy() {
        unbindService(mConexion)
        super.onDestroy()
    }
}